source/mongodb-handshake/handshake.rst

```diff
@@ -224,6 +224,70 @@ Example::
         - "Oracle JVM EE 9.1.1"
 
 
+--------------------------
+Speculative Authentication
```

Comment:
Do we need to add a note to https://github.com/mongodb/specifications/blob/master/source/command-monitoring/command-monitoring.rst#security to ensure that drivers redact `speculativeAuthenticate` and `speculativeSaslStart`? Actually I believe that the initial isMaster handshake is not supposed to be published at all but I can't find that requirement anywhere.  Should we add a note to this spec to say that the initial handshake MUST NOT be published?

Reply:
Good point. [Command Monitoring](https://github.com/mongodb/specifications/blob/master/source/command-monitoring/command-monitoring.rst) does not say anything about not publishing `isMaster` commands and I don't see any mention of that in [SDAM](https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst), either.

I agree that we can discuss security concerns here and would also propose that we update [Command Monitoring: Security](https://github.com/mongodb/specifications/blob/master/source/command-monitoring/command-monitoring.rst#security) to include `isMaster` with `speculativeAuthenticate` or `speculativeSaslStart` in its table of prohibited commands.

Additionally, I think we can make a separate SPEC ticket to update Command Monitoring and/or SDAM to say that `isMaster` commands shouldn't be reported with APM. I don't think any drivers actually do that, and it seems redundant given that SDAM has its own monitoring events.

Reply:
I have updated the Security section in Command Monitoring to include `isMaster` with `speculativeAuthenticate`.

Reply:
I feel like this is going to be painful to implement, since it's the only redacted command that requires looking at the content of the command rather than just the name.  I think we should go ahead and make @jmikola 's suggested change about not reporting ismaster commands at all.  Except that this is not only about SDAM.  It's about the handshake for all new connections.  So we could say that any command executed as part of the handshake MUST NOT be reported via command monitoring.

Reply:
I don't think it's _technically_ difficult to implement, but I would agree that it might be odd to conditionally exclude commands from APM based on what they contain. If a user was expecting to see handshake or monitoring commands today, they might wonder why those suddenly disappear under certain circumstances.

I think excluding all handshake stuff from APM is a reasonable compromise, but that seems like an independent task that would be a dependency of this SPEC ticket. Tests for not logging handshakes with APM wouldn't necessarily involve `speculativeAuthenticate` at all, so I expect that would warrant its own set of tests (assuming we have tests for handshakes today -- I'm not sure).

Do we want to roll that change into this SPEC ticket, with tests, or offload to a separate ticket for visibility? If we consolidate the change into this ticket, we'll definitely want to revise the corresponding DRIVERS ticket to inform folks of this extra change once everything is merged.

Reply:
@jyemin: In the interest of expediency, can we compromise by leaving the present text to conditionally exclude `isMaster` with `speculativeAuthenticate` from APM for now and then push "exclude all handshakes from APM" to a post-4.4 SPEC ticket?

Reply:
It is technically difficult in some drivers (Java, .NET) because of the way they do wire serialization.  That said, I think both of those drivers already exclude handshake-related commands from command monitoring, so this requirement becomes a no-op.

So I'm good with your suggestion.

Reply:
@ShaneHarvey ok to resolve this thread?